pkg/kubelet/cri/remote/remote_runtime.go

```diff
@@ -65,21 +67,102 @@ func NewRemoteRuntimeService(endpoint string, connectionTimeout time.Duration) (
 		return nil, err
 	}
 
-	return &remoteRuntimeService{
-		timeout:       connectionTimeout,
-		runtimeClient: runtimeapi.NewRuntimeServiceClient(conn),
-		logReduction:  logreduction.NewLogReduction(identicalErrorDelay),
-	}, nil
+	service := &remoteRuntimeService{
+		timeout:      connectionTimeout,
+		logReduction: logreduction.NewLogReduction(identicalErrorDelay),
+	}
+
+	if err := service.determineAPIVersion(conn, apiVersion); err != nil {
+		return nil, err
+	}
+
+	return service, nil
+}
+
+func (r *remoteRuntimeService) determineAPIVersion(conn *grpc.ClientConn, apiVersion internalapi.APIVersion) error {
+	ctx, cancel := getContextWithTimeout(r.timeout)
+	defer cancel()
+
+	switch apiVersion {
+	case internalapi.APIVersionV1:
+		klog.V(2).InfoS("Using selected CRI v1 API")
+		r.runtimeClientV1 = runtimeapiV1.NewRuntimeServiceClient(conn)
+
+	case internalapi.APIVersionV1alpha2:
+		klog.V(2).InfoS("Using selected CRI v1alpha2 API")
+		r.runtimeClientV1alpha2 = runtimeapiV1alpha2.NewRuntimeServiceClient(conn)
+
+	case "": // Automatically determine the API version while preferring V1
+		klog.V(4).InfoS("Finding the CRI API version")
+		r.runtimeClientV1 = runtimeapiV1.NewRuntimeServiceClient(conn)
+
+		if _, err := r.runtimeClientV1.Version(ctx, &runtimeapiV1.VersionRequest{}); err == nil {
+			klog.V(2).InfoS("Using CRI v1 API")
+
+		} else if status.Code(err) == codes.Unimplemented {
+			klog.V(2).InfoS("Falling back to CRI v1alpha2 API (deprecated)")
+			r.runtimeClientV1alpha2 = runtimeapiV1alpha2.NewRuntimeServiceClient(conn)
```

Comment:
The gRPC can re-dial if CRI restarts. If the CRI upgrades the version to V1, the kubelet still uses `v1alpha2` protocol. Currently, both containerd and crio can support two protocols. If the other CRI implementation only supports one protocol, the kubelet needs to restart in this case.

Reply:
Interesting .. If the container runtime upgrades to a version supporting v1.. this instance of kubelet will be locked into the prior version of the api, and I think that is ok for this instance. If the container runtime upgrade does not support the version initially used the re-dial will fail.  I think we should at least manually test cases. I think it would be fair to add documentation notes to the CRI v1 changes going into 1.23 regarding these administrative considerations.

Reply:
@fuweid @mikebrow would it be fine if we state:

GRPC redial will always use the initially selected (or automatically determined) CRI API verison?

Reply:
yes.  And I think it's important to add that if the redial was due to the container runtime being upgraded that the container runtime must also support the initially selected version or the redial is expected to fail, requiring a restart of kubelet.

Reply:
Added this information to the release note as well as the documentation of the `determineAPIVersion` method.